### PR TITLE
Develop authorization

### PR DESCRIPTION
**Pull request by kacpersmal**

### CODE REVIEW

File: src/controllers/auth/Authorize.controller.ts [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  email: Joi.string()
+    .required()
+    .email({ tlds: { allow: false } }),
+  password: Joi.string().required(),
+});
+
+const AuthorizeController = async (req: Request, res: Response, next: NextFunction) => {
+  const user = await GetUserByEmail(req.body.email);
+
+  if (user == undefined) return next(InvalidCredentialsException);
```
Comment (Miszczu2137) [inline src/controllers/auth/Authorize.controller.ts:20]: Remember about `{}`

File: src/controllers/user/DeleteUser.controller.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  userId: Joi.string().required(),
+});
+
+const UserNotFoundException = (id: string) => new HttpException(StatusCode.NOT_FOUND, 'User with given id does not exist!', { id: id });
+const UnpriviligedAccessException = () => new HttpException(StatusCode.FORBIDDEN, 'Unpriviliged Access');
+
+const DeleteUserController = async (req: Request, res: Response, next: NextFunction) => {
+  const id = req.params['userId'];
+
+  // console.log(res.locals.id != id || res.locals.user_role != 'admin');
```
Comment (Miszczu2137) [inline src/controllers/user/DeleteUser.controller.ts:18]: Remember to remove all unnecessary code.

File: src/controllers/user/EditUser.controller.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+});
+
+const UserNotFoundException = new HttpException(StatusCode.NOT_FOUND, 'User with given id does not exists');
+
+const EditUserController = async (req: Request, res: Response, next: NextFunction) => {
+  const userId = req.params['userId'];
+  const user = await GetUserById(userId);
+  if (user == undefined) return next(UserNotFoundException);
+
+  await SaveUser(userId, { email: req.body.email, password_hash: await HashPassword(req.body.password) });
```
Comment (Miszczu2137) [inline src/controllers/user/EditUser.controller.ts:20]: During edition email and password also should be validating.

File: src/middleware/Guards/AdminGuard.ts [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+import StatusCode from '../../utils/StatusCodes';
+import jwt from 'jsonwebtoken';
+import config from 'config';
+const AdminGuard = (req: Request, res: Response, next: NextFunction) => {
+  const token = GetTokenFromRequest(req);
+  if (!token) return res.sendStatus(StatusCode.UNAUTHORIZED);
+
+  jwt.verify(token, config.get('JWT.key'), (err: any, payload: any) => {
+    if (err) return res.sendStatus(StatusCode.UNAUTHORIZED);
+    if (payload.user_role != 'admin') return res.sendStatus(StatusCode.FORBIDDEN);
```
Comment (Miszczu2137) [inline src/middleware/Guards/AdminGuard.ts:12]: Roles shouldn't be hardcoded like that. Please use some constants for it.

Comment (kacpersmal) [reply src/middleware/Guards/AdminGuard.ts:12]: sikved

File: src/controllers/user/EditUser.controller.ts (hunk 3 above)
Comment (kacpersmal) [reply src/controllers/user/EditUser.controller.ts:20]: solved

File: src/controllers/user/DeleteUser.controller.ts (hunk 2 above)
Comment (kacpersmal) [reply src/controllers/user/DeleteUser.controller.ts:18]: solved

File: src/controllers/auth/Authorize.controller.ts (hunk 1 above)
Comment (kacpersmal) [reply src/controllers/auth/Authorize.controller.ts:20]: solved

**Review by Miszczu2137 — APPROVED** [review]

**Pull request merged**